### PR TITLE
Support connection migration on the server-side

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should be able to support connection migration the server side. This can be tested via quiche:

```
cargo run --bin quiche-client -- --no-verify --perform-migration https://127.0.0.1:9999/`
```

Modifications:

Be able to support migrations by offering the remote peer connection ids to use.

Result:

Be able to do connection migration in all cases.

### CODE REVIEW

File: codec-native-quic/pom.xml [hunk 1]
```diff
@@ -54,9 +54,9 @@
     <quicheHomeDir>${project.build.directory}/quiche</quicheHomeDir>
     <quicheHomeBuildDir>${quicheHomeDir}/build</quicheHomeBuildDir>
     <quicheHomeIncludeDir>${quicheHomeDir}/quiche/include</quicheHomeIncludeDir>
-    <quicheRepository>https://github.com/cloudflare/quiche</quicheRepository>
-    <quicheBranch>master</quicheBranch>
-    <quicheCommitSha>af368e9287ea975d184f9f66df52dbf109adf0e4</quicheCommitSha>
+    <quicheRepository>https://github.com/normanmaurer/quiche</quicheRepository>
+    <quicheBranch>c_api</quicheBranch>
+    <quicheCommitSha>b140659a16bba84d2a30d8990ec5c9c9b0b88104</quicheCommitSha>
```
Comment (normanmaurer) [inline codec-native-quic/pom.xml:59]: This change should be reverted once quiche merged the required pr:

https://github.com/cloudflare/quiche/pull/1656

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicChannel.java [hunk 2]
```diff
@@ -864,6 +865,62 @@ void recv(InetSocketAddress recipient, InetSocketAddress sender, ByteBuf buffer)
...
+            if (left > 0) {
+                boolean sendAndFlush = false;
+                ByteBuf connIdBuffer = alloc().directBuffer(key.remaining());
+                try {
+                    do {
+                        ByteBuffer key = key().duplicate();
+                        ByteBuffer srcId = connectionIdGenerator.newId(key, key.remaining());
+                        connIdBuffer.clear();
+                        connIdBuffer.writeBytes(srcId.duplicate());
+                        // TODO: Generate the reset token somehow. Maybe allow the user to plugin how this is done ?
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicChannel.java:901]: I need to think a bit more about this before we merge this

**Comment by normanmaurer** [comment]
There is still an issue in the code that can cause stale entries in the connection map. Working on it

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/Quiche.java [hunk 3]
```diff
@@ -547,6 +547,12 @@ static native int quiche_conn_stream_priority(
      */
     static native int quiche_conn_max_send_udp_payload_size(long connAddr);
 
+    static native int quiche_conn_source_cids_left(long connAddr);
+
+    static native long quiche_conn_new_source_cid(long connAddr, long scidAddr, int scidLen, byte[] resetToken, boolean retire_if_needed);
+
+    static native byte[] quiche_conn_retired_scid_next(long connAddr);
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/Quiche.java:554]: will add some javadocs here as well

**Review by bryce-anderson — APPROVED** [review]
I'm not super familiar with this but from the surface it seems reasonable to me.

**Pull request merged**